src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs

```diff
@@ -1716,6 +1734,35 @@ internal static bool IsSpanOrListType(CSharpCompilation compilation, TypeSymbol
             elementType = default;
             return false;
         }
+
+        internal static bool IsKeyValuePairType(CSharpCompilation compilation, TypeSymbol? targetType, WellKnownType keyValuePairType, out TypeWithAnnotations keyType, out TypeWithAnnotations valueType)
+        {
+            if (targetType is NamedTypeSymbol { Arity: 2 } namedType
+                && ReferenceEquals(namedType.OriginalDefinition, compilation.GetWellKnownType(keyValuePairType)))
+            {
+                var typeArguments = namedType.TypeArgumentsWithAnnotationsNoUseSiteDiagnostics;
+                keyType = typeArguments[0];
+                valueType = typeArguments[1];
+                return true;
+            }
+            keyType = default;
+            valueType = default;
+            return false;
+        }
+
+        internal static bool CollectionUsesKeyValuePairs(CSharpCompilation compilation, CollectionExpressionTypeKind collectionTypeKind, TypeSymbol elementType, out TypeWithAnnotations keyType, out TypeWithAnnotations valueType)
```

Comment:
>CollectionUsesKeyValuePairs [](http://example.com/codeflow?start=29&length=27)

Since a dictionary type is defined as a collection type whose element type is a KVP, we could name this helper `IsDictionaryType`